Added NetApp OnCommand Cloud Manager provider

miechus opened the pull request:
NetApp OCCM is used by enterprises to implement scalable cloud-based NAS solutions. This provider adds support for managing OCCM volumes. Please note that in order to run acceptance tests it is required that the OCCM environment is set up and configured (both VSA (non-HA) and HA environments are required and they need to have the Active Directory set up in order for the tests to pass).

Depending on the OCCM processing speed the "*_data_change" tests may fail due to the fact that OCCM sometimes takes a significant amount of time to change underlying volume type.

miechus commented:
I reordered the provider list to be more alphabetical, please ignore the change as needed.

catsby commented:
Hello! 

My name is Clint, I'm a member of the Terraform team @ HashiCorp.

I’ve taken a look at the provider here and would like to say great work so far! I do have feedback outlined below that I’d like to see addressed before we move on to the next steps. I’m opening this issues as a sort of checklist for tracking remaining items and discussion. The review was done on the git commit https://github.com/hashicorp/terraform/pull/15794/commits/1601495f1146d3b7fa6a2438e7fd1ba2564ddbfb .

- [ ] `resource_cloud_volume`: `create_aggregate_if_not_found` attribute either needs a default value or use the new [GetOkExists](https://github.com/hashicorp/terraform/blob/46af0c94931dbd58ce6a7560f6ffdee59bb08d38/helper/schema/resource_data.go#L117) helper method. On line 
https://github.com/hashicorp/terraform/pull/15794/files#diff-3c52100eeaa795cd2528be59e4e528c7R520 , without `GetOkExists` or a default value, a user cannot send `false` here. Also, this attribute is not mentioned in the `Update` method, so once set a user still cannot modify that value. If this is intended behavior, then it should be documented 
- [ ] `resource_cloud_volume`: `max_num_disks_approved_to_add` is optional with on default, but a default will be used from the results of an API call if not provided. A default should be added, or this attribute should be `Computed`. Also, `max_num_disks_approved_to_add` doesn’t seem to be re-read in the `READ` method, so a user may not know what value was used here
- [ ] Please confirm all attributes that can be updated are included in the `UPDATE` method, as well as read back and saved to state in the `READ` method
- [ ] `resource_cloud_volume`: it seems the `resourceCloudVolumeDelete` method will be stuck in a “failed” cycle if the resource is destroyed outside of the given Terraform project, meaning if it’s destroyed via a web console, it seems the `WaitForRequst` method will error if it gets s 404. In this situation you need to check if the error is “not found”, which in the context of trying to destroy a resource, should simply log the not found message, call `d.SetId(“”)` and return nil. This applies to most all `DELETE` methods

In general this looks to be in good shape, but there are things above that need to be addressed. As you probably know, Terraform now live in the [terraform-providers](https://github.com/terraform-providers) organization, and not in the main [hashicorp/terraform](https://github.com/hashicorp/terraform) repository. For this addition, we’d like just be able to move this ourselves if you can resolve the issues above. Also, we have a more formal process outlined for adding Providers to `terraform-providers`, and we’ll need some mutual agreement on maintenance, testing accounts, and such. Please see the outline here:

- https://www.terraform.io/guides/terraform-provider-development-program.html

Let us know if you have any questions! 

Cheers,
Clint

johnpeak commented:
Hi Clint,

We’re very appreciative of you taking the time to review this!  Thank you.

I’m looping in Mac, who did the actual work.  He can provide you answers to your questions.

thx!
john


From: Clint
Reply-To: hashicorp/terraform
Date: Wednesday, September 27, 2017 at 12:24 PM
To: hashicorp/terraform
Cc: John Peak, Manual
Subject: Re: [hashicorp/terraform] Added NetApp OnCommand Cloud Manager provider (#15794)


Hello!

My name is Clint, I'm a member of the Terraform team @ HashiCorp.

I’ve taken a look at the provider here and would like to say great work so far! I do have feedback outlined below that I’d like to see addressed before we move on to the next steps. I’m opening this issues as a sort of checklist for tracking remaining items and discussion. The review was done on the git commit 1601495<https://github.com/hashicorp/terraform/commit/1601495f1146d3b7fa6a2438e7fd1ba2564ddbfb> .

  *   resource_cloud_volume: create_aggregate_if_not_found attribute either needs a default value or use the new GetOkExists<https://github.com/hashicorp/terraform/blob/46af0c94931dbd58ce6a7560f6ffdee59bb08d38/helper/schema/resource_data.go#L117> helper method. On line
https://github.com/hashicorp/terraform/pull/15794/files#diff-3c52100eeaa795cd2528be59e4e528c7R520 , without GetOkExists or a default value, a user cannot send false here. Also, this attribute is not mentioned in the Update method, so once set a user still cannot modify that value. If this is intended behavior, then it should be documented
  *   resource_cloud_volume: max_num_disks_approved_to_add is optional with on default, but a default will be used from the results of an API call if not provided. A default should be added, or this attribute should be Computed. Also, max_num_disks_approved_to_add doesn’t seem to be re-read in the READ method, so a user may not know what value was used here
  *    Please confirm all attributes that can be updated are included in the UPDATE method, as well as read back and saved to state in the READ method
  *   resource_cloud_volume: it seems the resourceCloudVolumeDelete method will be stuck in a “failed” cycle if the resource is destroyed outside of the given Terraform project, meaning if it’s destroyed via a web console, it seems the WaitForRequst method will error if it gets s 404. In this situation you need to check if the error is “not found”, which in the context of trying to destroy a resource, should simply log the not found message, call d.SetId(“”) and return nil. This applies to most all DELETE methods

In general this looks to be in good shape, but there are things above that need to be addressed. As you probably know, Terraform now live in the terraform-providers<https://github.com/terraform-providers> organization, and not in the main hashicorp/terraform<https://github.com/hashicorp/terraform> repository. For this addition, we’d like just be able to move this ourselves if you can resolve the issues above. Also, we have a more formal process outlined for adding Providers to terraform-providers, and we’ll need some mutual agreement on maintenance, testing accounts, and such. Please see the outline here:

  *   https://www.terraform.io/guides/terraform-provider-development-program.html

Let us know if you have any questions!

Cheers,
Clint

—
You are receiving this because you are subscribed to this thread.
Reply to this email directly, view it on GitHub<https://github.com/hashicorp/terraform/pull/15794#issuecomment-332577268>, or mute the thread<https://github.com/notifications/unsubscribe-auth/ADwtEw70B_saa9AHLr1_NIJizX5jdQwdks5smnavgaJpZM4O1vwo>.

CONFIDENTIALITY  NOTICE: This message is intended for the use of the individual or entity to which it is addressed and may contain information that is confidential, privileged and exempt from disclosure under applicable law. If the reader of this message is not the intended recipient, you are hereby notified that any printing, copying, dissemination, distribution, disclosure or forwarding of this communication is strictly prohibited. If you have received this communication in error, please contact the sender immediately and delete it from your system. Thank You.

miechus commented:
Clint,

I'll be working on addressing your comments this week and will let you know when completed. I may have a few questions as some of the parameters are provided on input but cannot be retrieved during the read.

Thanks

Mac

catsby commented:
Sounds good, thanks!

catsby commented:
Hey Mac! Sorry for the delay. 

> The delete operation first attempts to refresh the state, which fails as the volume no longer exists. As a result, the delete logic never fires. How would you like me to reproduce the usecase you described?

Can you elaborate on what happens when you say "refresh the state, which fails"? If Terraform is throwing an error and saying "this *thing* no longer exists", we need to address that. For example, in `resourceCloudVolumeRead`, if the volume is not found, I assume there is an error returned? We want to check that error, and if it's a 404 or otherwise known "not found" error, we need to do something like this:

```go
// pseudocode
if err == ErrNotFound {
  log.Printf("[WARN] Volume (%s) not found, removing from state", id)
  d.SetId("")
  return nil
}
```

This will remove the volume from state. The same should be done for `resourceCloudVolumeDelete`. There's actually an additional method you can implement in the resource schema, `Exists`, alongside the `Read` and `Delete` methods that does basically this. See the [AWS KMS Key Resource](https://github.com/terraform-providers/terraform-provider-aws/blob/befa952c33cdf4b391a7b6951d4ff04fd672bcfd/aws/resource_aws_kms_key.go#L22) for an example. 

The idea here is if a resource is deleted outside of Terraform, we don't want users stuck with a "resource not found" error. If it's not there, then a `terraform plan` should show a creation event in the plan. A `terraform refresh` should log the 404 remove it (this could happen in `Exists` or `Read`). A `terraform delete` should log that it wasn't there and set `d.SetId("")` and return nil. 

`Exists` is not a require method but it's probably recommended. If your resource defines `Exists`, Terraform will call it before `Read` and you won't need to do these kinds of checks in `Read` or `Delete`. That said, if it does exist, you'll likely do a very similar and maybe even duplicate API call in `Read`, et. al, so if number of API calls are of concern, then you should just do the checks in-line with `Read` or `Delete`.


>  I'm not sure how we can address this issue - it is an edge case that will rarely show up in the field.

I truly hope it is an edge case rarely encountered, but unfortunately you'd be surprised how many users hit edge cases 😄 

If possible polling should be done to retry or monitor calls until they are successful. I'll leave it to you to decide that. 

I think I heard from Burzin that there are credentials provided, but please let me know what you think of my suggestions. I’ve checked off the 3 completed items. I’ll wait to hear back from you on my suggestions, when I do I’ll double check the acceptance tests with the credentials you’ve provided.

Thanks!

catsby commented:
Hey all – as a reminder, I mentioned in my initial review that this needs to be moved into it's own repository to really move forward. The structure can be mimicked from the AWS or Fastly provider as examples:

- https://github.com/terraform-providers/terraform-provider-aws
- https://github.com/terraform-providers/terraform-provider-fastly

Thanks!

miechus commented:
@catsby Clint, I created a separate repo containing just the provider here:
https://github.com/miechus/terraform-provider-netapp

I added resource busy checks and the resource exists checks. Please let me know if there's anything else I should change.

Thanks

Mac

catsby commented:
Going to close this in favor of reviewing and discussing over at https://github.com/miechus/terraform-provider-netapp
